appcues/src/test/java/com/appcues/analytics/AutoPropertyDecoratorTest.kt

```diff
@@ -0,0 +1,201 @@
+            name = AnalyticsEvent.ScreenView.eventName,
+            attributes = hashMapOf(
+                ActivityRequestBuilder.SCREEN_TITLE_ATTRIBUTE to "Screen 1"
+            )
+        )
+        val event2 = EventRequest(
+            name = AnalyticsEvent.ScreenView.eventName,
+            attributes = hashMapOf(
+                ActivityRequestBuilder.SCREEN_TITLE_ATTRIBUTE to "Screen 2"
+            )
+        )
+        // when
+        with(autoPropertyDecorator.decorateTrack(event)) {
+            // then
+            assertThat(attributes.containsKey("_identity")).isTrue()
+            with(attributes["_identity"] as Map<String, Any>) {
+                assertThat(containsKey("_lastScreenTitle")).isFalse()
+                assertThat(get("_currentScreenTitle")).isEqualTo("Screen 1")
+                assertThat(get("_sessionPageviews")).isEqualTo(1)
+            }
+        }
+
+        with(autoPropertyDecorator.decorateTrack(event2)) {
+            // then
+            assertThat(attributes.containsKey("_identity")).isTrue()
+            with(attributes["_identity"] as Map<String, Any>) {
+                assertThat(get("_lastScreenTitle")).isEqualTo("Screen 1")
+                assertThat(get("_currentScreenTitle")).isEqualTo("Screen 2")
+                assertThat(get("_sessionPageviews")).isEqualTo(2)
+            }
+        }
+    }
+
+    @Test
+    fun `decorateTrack SHOULD decorate identity properties WHEN event is SessionStarted`() {
+        // given
+        val event = EventRequest(
+            name = AnalyticsEvent.SessionStarted.eventName,
+        )
+        // when
+        with(autoPropertyDecorator.decorateTrack(event)) {
+            // then
+            with(attributes["_identity"] as Map<String, Any>) {
+                assertThat(containsKey("_lastScreenTitle")).isFalse()
+                assertThat(containsKey("_currentScreenTitle")).isFalse()
+                assertThat(get("_sessionPageviews")).isEqualTo(0)
+                assertThat(get("_sessionRandomizer")).isNotEqualTo(0)
+            }
+        }
+    }
+
+    @Test
+    fun `decorateTrack SHOULD decorate identity WITH new sessionRandomizer WHEN event is SessionStarted`() {
+        // given
+        val event = EventRequest(
+            name = AnalyticsEvent.SessionStarted.eventName,
+        )
+        var sessionRandomizer = 0
+        // when
+        with(autoPropertyDecorator.decorateTrack(event)) {
+            // then
+            with(attributes["_identity"] as Map<String, Any>) {
+                assertThat(get("_sessionRandomizer")).isNotEqualTo(sessionRandomizer)
+                sessionRandomizer = get("_sessionRandomizer") as Int
+            }
+        }
+        // when
+        with(autoPropertyDecorator.decorateTrack(event)) {
+            // then
+            with(attributes["_identity"] as Map<String, Any>) {
+                assertThat(get("_sessionRandomizer")).isNotEqualTo(sessionRandomizer)
```

Comment:
suppose there is a very slim chance that this test will randomly fail?

Reply:
okay so I did some refactoring, please take a look and see if it makes sense.

I split your rule into two rules, this is how I used in previous jobs too, so they work basically as small "plugins" to test files.


sessonRandomizer test is a tricky one, let me know if you have a better idea how we can do it.

Reply:
I thought i replied on this one but maybe it didnt save.  Anyway - splitting into two Rules seems all good to me! Its originally how I had it actually, just combined them since I was always using both.  But if there is a use case for using individually then it makes sense to make the more granular.